src/test/java/io/jenkins/plugins/analysis/warnings/FlexiblePublishITest.java

```diff
@@ -0,0 +1,159 @@
+import io.jenkins.plugins.analysis.warnings.recorder.pageobj.DetailsTab;
+import io.jenkins.plugins.analysis.warnings.recorder.pageobj.DetailsTab.TabType;
+import io.jenkins.plugins.analysis.warnings.recorder.pageobj.IssueRow;
+import io.jenkins.plugins.analysis.warnings.recorder.pageobj.IssuesTable;
+
+import static org.assertj.core.api.Assertions.*;
+
+/**
+ * Test the flexible publish plugin in combination with the warnings-ng-plugin.
+ *
+ * @author Tobias Redl
+ * @author Andreas Neumeier
+ */
+public class FlexiblePublishITest extends IntegrationTestWithJenkinsPerSuite {
+    private static final String JAVA_FILE = "java2Warnings.txt";
+    private static final String JAVA_FILE2 = "java-start.txt";
+    private static final String JAVA_PATTERN = "**/*.txt";
+    private static final String TOOL_ID = "sampleId1";
+    private static final String TOOL_ID2 = "sampleId2";
+
+    /**
+     * Test that the same tool can be used twice with different configuration.
+     */
+    @Test
+    public void shouldAnalyseJavaTwiceWithHealthReport() {
+        FreeStyleProject project = createFreeStyleProject();
+        copySingleFileToWorkspace(project, JAVA_FILE, JAVA_FILE);
+        copySingleFileToWorkspace(project, JAVA_FILE2, JAVA_FILE2);
+
+        IssuesRecorder publisher = constructJavaIssuesRecorder(JAVA_FILE, TOOL_ID, 1, 9);
+        IssuesRecorder publisher2 = constructJavaIssuesRecorder(JAVA_PATTERN, TOOL_ID2, 1, 3);
+
+        project.getPublishersList().add(new FlexiblePublisher(Arrays.asList(
+                constructConditionalPublisher(publisher),
+                constructConditionalPublisher(publisher2)
+        )));
+
+        hudson.model.Run<?, ?> run = buildSuccessfully(project);
+        List<AnalysisResult> results = getAnalysisResults(run);
+        assertThat(results).hasSize(2);
+        assertThat(results.get(0).getId()).isEqualTo(TOOL_ID);
+        assertThat(results.get(1).getId()).isEqualTo(TOOL_ID2);
+        List<HealthReport> healthReports = project.getBuildHealthReports();
+        assertThat(healthReports.get(0).getScore()).isEqualTo(0);
+        assertThat(healthReports.get(1).getScore()).isEqualTo(80);
+        assertThat(healthReports.get(2).getScore()).isEqualTo(100);
+        checkDetailsViewForIssues(project, results.get(0), results.get(0).getId(), 2);
+        checkDetailsViewForIssues(project, results.get(1), results.get(1).getId(), 4);
+    }
+
+    /**
+     * Test that two java issue recorder can run with different configuration (one with issue filter, one without).
+     */
+    @Test
+    public void shouldAnalyseJavaTwiceWithOneIssueFilter() {
+        FreeStyleProject project = createFreeStyleProject();
+        copySingleFileToWorkspace(project, JAVA_FILE, JAVA_FILE);
+        copySingleFileToWorkspace(project, JAVA_FILE2, JAVA_FILE2);
+
+        IssuesRecorder publisher = constructJavaIssuesRecorder(JAVA_PATTERN, TOOL_ID);
+
+        RegexpFilter filter = new ExcludeFile(".*File.java");
```

Comment:
line91-93: extract a createFilter method so the abstraction level is the same for all statements.